src/typing/fields.ml

```diff
@@ -239,297 +239,266 @@ let field_access ctx mode f famode e p =
-			begin match e.eexpr with
-				| TConst TSuper -> raise Not_found
-				| _ -> using_field ctx mode e i p
-			end
-		with Not_found -> try
-			loop_dyn c params
-		with Not_found -> try
-			(* if we have an abstract constraint we have to check its static fields and recurse (issue #2343) *)
-			begin match c.cl_kind with
+	let rec type_field_by_list f = function
+		| [] -> raise Not_found
+		| x :: l -> try f x with Not_found -> type_field_by_list f l
+	in
+	let type_field_by_forward f meta a =
+		let _,el,_ = Meta.get meta a.a_meta in
+		if el <> [] && not (List.exists (fun e -> match fst e with
+			| EConst (Ident i' | String (i',_)) -> i' = i
+			| _ -> error "Identifier or string expected as argument to @:forward" (pos e)
+		) el) then raise Not_found;
+		f()
+	in
+	let type_field_by_forward_static f a =
+		type_field_by_forward f Meta.ForwardStatics a
+	in
+	let type_field_by_forward_member f e a tl =
+		let f () = f { e with etype = Abstract.get_underlying_type ~return_first:true a tl } in
+		type_field_by_forward f Meta.Forward a
+	in
+	let rec type_field_by_type ?t e =
+		let t = get_t e t in
+		let field_access f fmode = field_access ctx mode f fmode e p in
+		match t with
+		| TInst (c,tl) ->
+			(try
+				let c2, t, f = class_field ctx c tl i p in
+				let fmode = match c2 with None -> FHAnon | Some (c,tl) -> FHInstance (c,tl) in
+				field_access f fmode
+			with Not_found ->
+				match c.cl_kind with
 				| KTypeParameter tl ->
-					let rec loop tl = match tl with
-						| t :: tl ->
-							begin match follow t with
-								| TAbstract({a_impl = Some c},tl) when PMap.mem i c.cl_statics ->
-									let e = mk_cast e t p in
-									type_field cfg ctx e i p mode with_type;
-								| _ ->
-									loop tl
-							end
-						| [] ->
-							raise Not_found
-					in
-					loop tl
-				| _ ->
-					raise Not_found
-			end
-		with Not_found ->
-			if PMap.mem i c.cl_statics then error ("Cannot access static field " ^ i ^ " from a class instance") pfield;
-			no_field())
-	| TDynamic t ->
-		(try
-			using_field ctx mode e i p
-		with Not_found ->
-			AKExpr (mk (TField (e,FDynamic i)) t p))
-	| TAnon a ->
-		(try
-			let f = PMap.find i a.a_fields in
-			if has_class_field_flag f CfImpl && not (has_class_field_flag f CfEnum) then display_error ctx "Cannot access non-static abstract field statically" pfield;
-			begin match mode with
-			| MCall _ when has_class_field_flag f CfOverload ->
-				()
-			| _ ->
-				if not (has_class_field_flag f CfPublic) && not ctx.untyped then begin
-					match !(a.a_status) with
+					type_field_by_list (fun t -> match follow t with
+						| TAbstract _ -> type_field_by_type (mk_cast e t p);
+						| _ -> raise Not_found
+					) tl
+				| _ -> raise Not_found
+			)
+		| TAnon a ->
+			(try
+				let f = PMap.find i a.a_fields in
+				if has_class_field_flag f CfImpl && not (has_class_field_flag f CfEnum) then display_error ctx "Cannot access non-static abstract field statically" pfield;
```

Comment:
It's a `TAnon` case. I guess this check is out of place.

Reply:
`TAnon` can have `AbstractStatics` status, this is for this case.
Here is [a test](https://github.com/HaxeFoundation/haxe/blob/development/tests/misc/projects/Issue5940/Main.hx) that checks that this is in place.

Reply:
Then it probably should be moved to `match !(a.a_status)` below.